src/mame/yamaha/ymdx9.cpp

```diff
@@ -0,0 +1,449 @@
+
+	// External RAM.
+	// 2 * 2kb RAM1 IC19 M5M118P.
+	map(0x0800, 0x0fff).ram().share("ram1");
+	map(0x1000, 0x1800).ram().share("ram2");
+
+	// YM21290 EGS
+	map(0x1800, 0x18F3).w(FUNC(yamaha_dx9_state::egs_w));
+
+	// ROM.
+	map(0xc000, 0xffff).rom().region("program", 0);
+}
+
+
+/**
+ * yamaha_dx9_state::dx9
+ */
+void yamaha_dx9_state::dx9(machine_config &config)
+{
+	// Initialise the HD63B03RP CPU.
+	// This oscillator frequency comes from the service manual.
+	HD6303R(config, m_maincpu, 9.4265_MHz_XTAL / 2);
+	m_maincpu->set_addrmap(AS_PROGRAM, &yamaha_dx9_state::mem_map);
+
+	// Unlike the DX7 only IO port 1 is used.
+	// The direction flags of other ports are set, however they are never read, or written.
+	m_maincpu->in_p1_cb().set(FUNC(yamaha_dx9_state::p1_r));
+	m_maincpu->in_p2_cb().set(FUNC(yamaha_dx9_state::unmapped_r));
+	m_maincpu->in_p3_cb().set(FUNC(yamaha_dx9_state::unmapped_r));
+	m_maincpu->in_p4_cb().set(FUNC(yamaha_dx9_state::unmapped_r));
+
+	m_maincpu->out_p1_cb().set(FUNC(yamaha_dx9_state::p1_w));
+	m_maincpu->out_p2_cb().set(FUNC(yamaha_dx9_state::unmapped_w));
+	m_maincpu->out_p3_cb().set(FUNC(yamaha_dx9_state::unmapped_w));
+	m_maincpu->out_p4_cb().set(FUNC(yamaha_dx9_state::unmapped_w));
+
+	NVRAM(config, "ram1", nvram_device::DEFAULT_ALL_0);
+	NVRAM(config, "ram2", nvram_device::DEFAULT_ALL_0);
+
+	// Configure the ADC. The clock speed here is a guess.
+	M58990(config, m_adc, 8_MHz_XTAL / 16);
+
+	// ADC source 4 is the battery voltage. Set this input to always read 0x80.
+	// If the read value is below 0x6F, the firmware considers this a low battery voltage.
+	m_adc->in_callback<4>().set_constant(0x80);
+
+	// Configure MIDI.
+	auto &midiclock(CLOCK(config, "midiclock", 500_kHz_XTAL));
+	midiclock.signal_handler().set(FUNC(yamaha_dx9_state::midiclock_w));
+
+	MIDI_PORT(config, "mdin", midiin_slot, "midiin").rxd_handler().set(FUNC(yamaha_dx9_state::midi_r));
+
+	auto &mdout(MIDI_PORT(config, "mdout", midiout_slot, "midiout"));
+	m_maincpu->out_ser_tx_cb().set(mdout, FUNC(midi_port_device::write_txd));
+
+	// Configure the LCD screen.
+	screen_device &lcd_screen(SCREEN(config, "lcd_screen", SCREEN_TYPE_LCD));
+	lcd_screen.set_refresh_hz(60);
+	lcd_screen.set_vblank_time(ATTOSECONDS_IN_USEC(2500)); /* not accurate */
+	lcd_screen.set_screen_update("lcdc", FUNC(hd44780_device::screen_update));
+	lcd_screen.set_size(6*16+1, 10*2+1);
+	lcd_screen.set_visarea_full();
+	lcd_screen.set_palette("palette");
+
+	PALETTE(config, "palette", FUNC(yamaha_dx9_state::palette_init), 3);
+
+	hd44780_device &lcdc(HD44780(config, "lcdc", 0));
+	lcdc.set_lcd_size(2, 16);
+	lcdc.set_pixel_update_cb(FUNC(yamaha_dx9_state::lcd_pixel_update));
+
+	config.set_default_layout(layout_dx9);
+}
+
+
+/**
+ * yamaha_dx9_state::unmapped_r
+ */
+uint8_t yamaha_dx9_state::unmapped_r(offs_t offset)
+{
+	return 0xFF;
+}
+
+
+/**
+ * yamaha_dx9_state::key_switch_scan_driver_r
+ */
+uint8_t yamaha_dx9_state::key_switch_scan_driver_r(offs_t offset)
```

Comment:
This whole function seems really bloated. With an ioport_array it would boil down to one line:

 `return m_keys[m_key_select]->read();`

(and this can be put into a lambda function then)